### PR TITLE
detecting whether some well control events happens to a well

### PR DESCRIPTION
**Pull request by GitPaean**
When there is some events happen to a well, we use the control mode
 from the DECK, and update the WellState based on the new control model.
 Otherwise, we can use the control mode from the previous well state,
 and keep the values from the previous well state as an intial guess.

### CODE REVIEW

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 1]
```diff
@@ -228,10 +252,24 @@ namespace Opm
             }
         }
 
-        void resize(const Wells* wells, size_t numCells, const PhaseUsage& pu)
+        void resize(const Wells* wells, size_t numCells)
```
Comment (GitPaean) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: I am not familiar with the usage of this function, @andlaus , please comment if the change is not safe.

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp (hunk 1 above)
Comment (andlaus) [reply opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: I don't know anything here either. Maybe @atgeirr does?

Comment (GitPaean) [reply opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Okay. I thought this is a function you wrote. If not, never mind.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 2]
```diff
@@ -707,13 +740,13 @@ namespace Opm
         }
 
 
-        bool isNewWell(const int w) const {
-            return is_new_well_[w];
+        bool effectiveEventsHappen(const int w) const {
+            return effective_events_happen_[w];
```
Comment (atgeirr) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Make sure this works if wells are shut or opened *during* a report step due to inoperable well or WTEST.

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 3]
```diff
@@ -707,13 +740,13 @@ namespace Opm
...
         }
 
 
-        void setNewWell(const int w, const bool is_new_well) {
-            is_new_well_[w] = is_new_well;
+        void setEffeciveEventHappen(const int w, const bool effective_events_happen) {
```
Comment (atgeirr) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Typo. Rename variable to `effective_event_occurred`, and functions to match.

File: opm/autodiff/StandardWell_impl.hpp [hunk 4]
```diff
@@ -1908,13 +1906,14 @@ namespace Opm
...
-            if ( !well_state.isNewWell(index_of_well_) ) {
+            if ( !well_state.effectiveEventsHappen(index_of_well_) ) {
                 for (int p = 0; p < np; ++p) {
                     // This is dangerous for new added well
                     // since we are not handling the initialization correctly for now
                     well_potentials[p] = well_state.wellRates()[index_of_well_ * np + p];
                 }
-            } else {
+            } else
+            {
```
Comment (atgeirr) [inline opm/autodiff/StandardWell_impl.hpp]: Why?

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp (hunk 2 above)
Comment (GitPaean) [reply opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Except this one, other comments were taken care of.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
I will check the opm-tests results locally and report it back, while it takes a little while to finish re-compiling. 

I am just update_data for later use.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=71 please

**Comment by GitPaean** [comment]
testing result of Model 2
With this PR
```
2324776 Total time (seconds):         3918.62
2324777 Solver time (seconds):        3890.34
2324778  Assembly time (seconds):     1264.22 (Failed: 128.216; 10.1419%)
2324779  Linear solve time (seconds): 2546.03 (Failed: 140.918; 5.53482%)
2324780  Update time (seconds):       52.6686 (Failed: 5.46468; 10.3756%)
2324781  Output write time (seconds): 30.7416
2324782 Overall Well Iterations:      729 (Failed: 24; 3.29218%)
2324783 Overall Linearizations:       2353 (Failed: 231; 9.81725%)
2324784 Overall Newton Iterations:    2023 (Failed: 231; 11.4187%)
2324785 Overall Linear Iterations:    51895 (Failed: 2487; 4.79237%)
```

With master branch
```
2316036 Total time (seconds):         4010.67
2316037 Solver time (seconds):        3983.43
2316038  Assembly time (seconds):     1241.27 (Failed: 116.335; 9.37223%)
2316039  Linear solve time (seconds): 2664.55 (Failed: 184.39; 6.92011%)
2316040  Update time (seconds):       50.5627 (Failed: 4.91004; 9.71079%)
2316041  Output write time (seconds): 29.6962
2316042 Overall Well Iterations:      736 (Failed: 31; 4.21196%)
2316043 Overall Linearizations:       2345 (Failed: 215; 9.16844%)
2316044 Overall Newton Iterations:    2015 (Failed: 215; 10.67%)
2316045 Overall Linear Iterations:    54281 (Failed: 3523; 6.4903%)
```


Testing result of Norne
With this PR
```
1286946 Total time (seconds):         960.075
1286947 Solver time (seconds):        954.086
1286948  Assembly time (seconds):     431.411 (Failed: 13.7957; 3.19782%)
1286949  Linear solve time (seconds): 493.969 (Failed: 12.1633; 2.46237%)
1286950  Update time (seconds):       15.973 (Failed: 0.547653; 3.42861%)
1286951  Output write time (seconds): 7.16553
1286952 Overall Well Iterations:      949 (Failed: 8; 0.842993%)
1286953 Overall Linearizations:       1980 (Failed: 63; 3.18182%)
1286954 Overall Newton Iterations:    1644 (Failed: 63; 3.83212%)
1286955 Overall Linear Iterations:    24959 (Failed: 550; 2.20361%)
```

With master branch
```
1286105 Total time (seconds):         994.779
1286106 Solver time (seconds):        988.526
1286107  Assembly time (seconds):     441.298 (Failed: 14.0591; 3.18585%)
1286108  Linear solve time (seconds): 518.326 (Failed: 12.8416; 2.4775%)
1286109  Update time (seconds):       15.9586 (Failed: 0.556411; 3.48659%)
1286110  Output write time (seconds): 7.48249
1286111 Overall Well Iterations:      950 (Failed: 8; 0.842105%)
1286112 Overall Linearizations:       1974 (Failed: 63; 3.19149%)
1286113 Overall Newton Iterations:    1638 (Failed: 63; 3.84615%)
1286114 Overall Linear Iterations:    25282 (Failed: 566; 2.23875%)
```

**Comment by atgeirr** [comment]
Great, Jenkins is green, I will merge this.

**Pull request merged**